python/src/equistore/operations/join.py

```diff
@@ -0,0 +1,236 @@
+            )
+            properties2 = np.hstack(
+                [np.ones(len(properties2), dtype=int).reshape(-1, 1), properties2]
+            )
+    else:
+        # if property_names are the same in each block and the values are
+        # unique we keep the names of the first block and stack the values
+        property_names = names1
+
+    properties_values = np.vstack([properties1, properties2])
+
+    return Labels(names=property_names, values=properties_values)
+
+
+def _join_blocks_along_properties(
+    block1: TensorBlock, block2: TensorBlock
+) -> TensorBlock:
+    if not np.all(block1.samples == block2.samples):
+        raise ValueError("`A` and `B` do not have the same samples.")
+
+    if block1.components != block2.components:
+        raise ValueError("`A` and `B` do not have the same components")
+
+    if sorted(block1.gradients_list()) != sorted(block2.gradients_list()):
+        raise ValueError("`A` and `B` do not have the same gradients")
+
+    properties = _join_labels(block1.properties, block2.properties)
+    result_block = TensorBlock(
+        values=_dispatch.hstack([block1.values, block2.values]),
+        samples=block2.samples,
+        components=[],
+        properties=properties,
+    )
+
+    for parameter, gradient1 in block1.gradients():
+        gradient2 = block2.gradient(parameter)
+
+        gradient_data = np.zeros(gradient1.data.shape[:-1] + (len(properties),))
+
+        # copy block1 gradient data over
+        gradient_data[..., : len(block1.properties)] = gradient1.data
+
+        # check where to put the block2 data
+        for i_grad, sample_grad in enumerate(gradient2.samples):
+            i_merged = gradient1.samples.position(sample_grad)
+            gradient_data[i_merged, ..., len(block1.properties) :] = gradient2.data[
+                i_grad
+            ]
+
+        result_block.add_gradient(
+            parameter, gradient_data, gradient1.samples, gradient1.components
+        )
+
+    return result_block
+
+
+def _join_blocks_along_samples(block1: TensorBlock, block2: TensorBlock) -> TensorBlock:
+    for property1, property2 in zip(block1.properties, block2.properties):
+        if tuple(property1) != tuple(property2):
+            raise ValueError("`A` and `B` do not have the same properties.")
+
+    if block1.components != block2.components:
+        raise ValueError("`A` and `B` do not have the same components")
+
+    if sorted(block1.gradients_list()) != sorted(block2.gradients_list()):
+        raise ValueError("`A` and `B` do not have the same gradients")
+
+    result_block = TensorBlock(
+        values=_dispatch.vstack([block1.values, block2.values]),
+        samples=_join_labels(block1.samples, block2.samples),
+        components=[],
+        properties=block1.properties,
+    )
+
+    for parameter, gradient1 in block1.gradients():
+        gradient2 = block2.gradient(parameter)
+
```

Comment:
this should also check that the gradients components are the same (they should, but nothing requires that a gradient wrt "positions" always has a given set of components.)

Reply:
same as above